ipld/get_namespaced_shares.go

```diff
@@ -0,0 +1,188 @@
+package ipld
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+
+	"github.com/gammazero/workerpool"
+	"github.com/ipfs/go-blockservice"
+	"github.com/ipfs/go-cid"
+	ipld "github.com/ipfs/go-ipld-format"
+
+	"github.com/celestiaorg/celestia-node/ipld/plugin"
+	"github.com/celestiaorg/nmt"
+	"github.com/celestiaorg/nmt/namespace"
+)
+
+// TODO(@distractedm1nd) Find a better figure than NumWorkersLimit for this pool.
+// Worker pool responsible for the goroutines spawned by getLeavesByNamespace
+var namespacePool = workerpool.New(NumWorkersLimit)
+
+// GetSharesByNamespace walks the tree of a given root and returns its shares within the given namespace.ID.
+func GetSharesByNamespace(
+	ctx context.Context,
+	bGetter blockservice.BlockGetter,
+	root cid.Cid,
+	nID namespace.ID,
+	maxShares int,
+) ([]Share, error) {
+	leaves, err := getLeavesByNamespace(ctx, bGetter, root, nID, maxShares)
+	if err != nil {
+		return nil, err
+	}
+
+	shares := make([]Share, len(leaves))
+	for i, leaf := range leaves {
+		shares[i] = leafToShare(leaf)
+	}
+
+	return shares, nil
+}
+
+// wrappedWaitGroup is needed because waitgroups do not expose their internal counter,
+// and we don't know in advance how many jobs we will have to wait for.
+type wrappedWaitGroup struct {
+	wg      sync.WaitGroup
+	jobs    chan *job
+	counter int64
+}
+
+func createWrappedWaitGroup(jobs chan *job) *wrappedWaitGroup {
+	wg := &wrappedWaitGroup{
+		wg:      sync.WaitGroup{},
+		jobs:    jobs,
+		counter: 0,
+	}
+	wg.Add(1)
+	return wg
+}
+
+func (w *wrappedWaitGroup) Add(count int64) {
+	w.wg.Add(int(count))
+	atomic.AddInt64(&w.counter, count)
+}
+
+func (w *wrappedWaitGroup) Done() {
+	w.wg.Done()
+	atomic.AddInt64(&w.counter, -1)
+	if atomic.LoadInt64(&w.counter) == 0 {
+		close(w.jobs)
+	}
+}
+
+func (w *wrappedWaitGroup) Wait() {
+	w.wg.Wait()
+}
+
+type fetchedBounds struct {
+	lowest  int64
+	highest int64
+}
+
+func (b *fetchedBounds) Update(index int64) {
+	for ok := false; !ok; {
+		ok = true
+		if lowest := atomic.LoadInt64(&b.lowest); index < lowest {
+			ok = atomic.CompareAndSwapInt64(&b.lowest, lowest, index)
+		}
+	}
+	// not an `else if` because an element can be both the lower and higher bound
+	// for example, if there is only one share in the namespace
+	for ok := false; !ok; {
+		ok = true
+		if highest := atomic.LoadInt64(&b.highest); index > highest {
+			ok = atomic.CompareAndSwapInt64(&b.highest, highest, index)
+		}
+	}
```

Comment:
```suggestion
	ok := false
	for !ok {
		if lowest := atomic.LoadInt64(&b.lowest); index < lowest {
			ok = atomic.CompareAndSwapInt64(&b.lowest, lowest, index)
		}
	}

	ok = false
	for !ok {
		if highest := atomic.LoadInt64(&b.highest); index > highest {
			ok = atomic.CompareAndSwapInt64(&b.highest, highest, index)
		}
	}
```
Same logic, while takes less cod and `ok` is allocated once.

Reply:
This would not work, because the if statement is not always entered

Reply:
Was my first attempt as well!

Reply:
Oh right!

Reply:
Ok, so the current code still feels overcomplicated, so I updated the suggestion which obsoleted the ok completely.

Reply:
@distractedm1nd, you might resolved it before seeing the update

Reply:
Updated again with the most succinct form of it. Now there is no other way to compress it even more